ggml-rpc.cpp

```diff
@@ -0,0 +1,1023 @@
+            }
+            case GET_ALIGNMENT: {
+                rpc_get_alignment(backend, output);
+                break;
+            }
+            case GET_MAX_SIZE: {
+                rpc_get_max_size(backend, output);
+                break;
+            }
+            case BUFFER_GET_BASE: {
+                rpc_buffer_get_base(input, output);
+                break;
+            }
+            case FREE_BUFFER: {
+                rpc_free_buffer(input);
+                break;
+            }
+            case BUFFER_CLEAR: {
+                rpc_buffer_clear(input);
+                break;
+            }
+            case SET_TENSOR: {
+                rpc_set_tensor(input);
+                break;
+            }
+            case GET_TENSOR: {
+                rpc_get_tensor(input, output);
+                break;
+            }
+            case COPY_TENSOR: {
+                rpc_copy_tensor(input, output);
+                break;
+            }
+            case GRAPH_COMPUTE: {
+                rpc_graph_compute(backend, input, output);
+                break;
+            }
+            case GET_DEVICE_MEMORY: {
+                // output serialization format: | free (8 bytes) | total (8 bytes) |
+                output.resize(2*sizeof(uint64_t), 0);
+                memcpy(output.data(), &free_mem, sizeof(free_mem));
+                memcpy(output.data() + sizeof(uint64_t), &total_mem, sizeof(total_mem));
+                break;
+            }
+            default: {
+                fprintf(stderr, "Unknown command: %d\n", cmd);
+                return;
+            }
+        }
+        uint64_t output_size = output.size();
+        if (!send_data(sockfd, &output_size, sizeof(output_size))) {
+            break;
+        }
+        if (!send_data(sockfd, output.data(), output_size)) {
+            break;
+        }
+    }
+}
+
+void start_rpc_server(ggml_backend_t backend, const char * endpoint, size_t free_mem, size_t total_mem) {
+    std::string host;
+    int port;
+    if (!parse_endpoint(endpoint, host, port)) {
+        return;
+    }
+#ifdef _WIN32
+    {
+        WSADATA wsaData;
+        int res = WSAStartup(MAKEWORD(2, 2), &wsaData);
+        if (res != 0) {
+            fprintf(stderr, "WSAStartup failed: %d\n", res);
+            return;
+        }
+    }
+#endif
+    auto server_socket = create_server_socket(host.c_str(), port);
+    if (server_socket == nullptr) {
+        fprintf(stderr, "Failed to create server socket\n");
+        return;
+    }
+    while (true) {
+        auto client_socket = socket_accept(server_socket->fd);
+        if (client_socket == nullptr) {
+            fprintf(stderr, "Failed to accept client connection\n");
+            return;
+        }
+        printf("Accepted client connection, free_mem=%zu, total_mem=%zu\n", free_mem, total_mem);
+        rpc_serve_client(backend, client_socket->fd, free_mem, total_mem);
+        printf("Client connection closed\n");
```

Comment:
Hi @rgerganov, thank your for this nice PR to add the distribution ability to llama.cpp. 
Used it on my machine for days, and found a problem here: 
  When the client close the connection, the server will not free the memory they allocated.

After dive into the source code, found that insteads of releasing the memory we just quit the inner loop and then immediately waiting for new connection.

So just wandering, should we track the `ALLOC_BUFFER` and `FREE_BUFFER`, then free the reminding buffers after client close?

Reply:
Update: create a patch for this on may fork, maybe you can have a look and give some advice: https://github.com/chraac/llama.cpp/tree/dev-fix-rpc-mem-alloc

Reply:
@chraac just a suggestion - I think for maintainers would be easier if you can open up a PR directly with the changes, so discussions goes directly there.

Reply:
> @chraac just a suggestion - I think for maintainers would be easier if you can open up a PR directly with the changes, so discussions goes directly there.

Yeah, thanks, already have a [PR ](https://github.com/ggerganov/llama.cpp/pull/7378) and have [some discussions](https://github.com/ggerganov/llama.cpp/issues/7407) for it.